CARDS-1847: Incorrect identification of upcoming visits with tokenless authentication when declining the Terms of Use

marta- opened the pull request:
[See Jira for description and steps to replicate](https://phenotips.atlassian.net/browse/CARDS-1847)

marta- committed:
CARDS-1847: Incorrect identification of upcoming visits with tokenless authentication under some circumstances

marta- committed:
CARDS-1847: Incorrect identification of upcoming visits with tokenless authentication under some circumstances
Comment update

sashaandjic commented:
@marta- it is still happening. I think once you click on the `BEGIN` button, it remembers the clinic and the next time goes directly to that one

https://user-images.githubusercontent.com/23243244/176764688-4cad6c2b-f665-4c2e-a4ce-385a18f2cf3f.mp4

.

sdumitriu commented:
This does not solve bug 2. Bug 2 is a race condition during logout:
- when logging out, multiple requests are sent: the logout itself, and load requests for `SubjectTypes/Patient.full.json` and `current form.deep.json`
- all 3 requests are sent with `Cookie: current visit subject token`
- the logout request comes back with a `clear cookie` header
- the other 2 come back with a `set cookie to the same visit subject token` header
- depending on which one of these 3 requests finishes first, the visit subject token may be still set after the logout

The real bug is that the other 2 requests should not be sent.

marta- commented:
> This does not solve bug 2. Bug 2 is a race condition during logout:

Looks like bug 2 is completely unrelated. I will move it to a separate task. @sashaandjic can you confirm this PR solves bug 1 at least?

sashaandjic commented:
BUG
Steps to reproduce:
- Decline UoT first
- Accept UoT when the litle dialog pops up
- Expected behavior: user should be redirected to select a clinic to fill in surveys; Actual behavior: nothing happens 
- Once user clicks on `REVIEW TERMS` bttn user can select the clinic
- Please note if user clicks on `DECLINE` bttn again, they will be redirected to the login page as expected

https://user-images.githubusercontent.com/23243244/177651347-85f704e6-416a-45ef-93a3-2bda160a58d9.mp4

marta- committed:
CARDS-1847: Incorrect identification of upcoming visits with tokenless authentication when declining the Terms of Use
Bug fix: clicking accept in the ToU confirmation dialog doesn't close it to reveal visit selection

marta- committed:
CARDS-1847: Incorrect identification of upcoming visits with tokenless authentication when declining the Terms of Use
Incidental bug fix: declining the Terms of Use clears the form but doesn't sign the patient out

marta- commented:
> Found a problem. Steps to reproduce:
> 
> * create a patient with 2 upcoming visits with surveys scheduled
> * identify as the patient for the first time
> * decline the Terms of Use
> * return to the identification page after declining - however the patient is not logged out
> * click on CONTINUE without filling in DOB and MRN (or HC)
> * current behavior: accept UoT dialog appears ---> expected behavior: prompt user to enter DOB and MRN (or HC)

Addressed with a quick fix in 120ef39. The page flickers a bit because of the page refresh, but I'm hoping this use case is rare enough that it won't bother anyone too much.

sashaandjic reviewed (approved):
tested
